### ISSUE TITLE
dlopen constraints

### ISSUE THREAD

**Issue by sdaly2107** [android/ndk#1143]
Hello

We have an issue where we end up with a deadlock when using dlopen - are there any constraints on calls within the context of dlopen (similar to limitations inside Windows dllmain)?

The sequence is:

-Test A runs and creates some files.
-Test B runs and calls dlopen with a test SO.
-The test SO has a global which does some initialisation.  At the same time, a thread wakes from the previous test and calls some of our internal library calls.  A C++ exception is thrown (rtti and fexceptions are enabled and usually work fine).  The catch block is never reached.  Because of this, our own access mutex is never released and the initialisation done under dlopen blocks forever.

If we create an Initialise function in the SO and call dlopen then initialise, all is good.  If we link using -export-dynamic and build on Windows, all is good, but this does not link under Linux.

What's interesting is that our other thread starts calling dl_iterate_phd which then tries to acquire a mutex (I'm guessing g_dl_mutex, which dlopen has already acquired).  Why does this thread start executing dl related functions?

NDK version: r20b
ABI: Intel x86_64
Build: Using NDK cmake toolchain on Windows and Linux

**Comment by DanAlbert** [comment]
A couple ideas:

> The catch block is never reached.

Never reached, or passed without catching the exception? https://android.googlesource.com/platform/ndk/+/master/docs/user/common_problems.md#rtti_exceptions-not-working-across-library-boundaries might be the culprit if it's just not being caught.

> If we create an Initialise function in the SO and call dlopen then initialise, all is good.

Static constructor execution order is undefined. If you're relying on the order of initialization, you need an explicit initialization function.

> Why does this thread start executing dl related functions?

Throwing an exception would call this, as could a number of other things. Hard to say without knowing which of your functions caused the call.

**Comment by sdaly2107** [comment]
The try catch is within a single function and the catch is never reached.  We tried changing to c++_shared and the same behaviour is observed.

A single object is used to initialise our logging in a controlled manner.

OK, so looks like as part of the unwind after throwing, dl_iterate_phd  is called which tries to lock g_dl_mutex.  

I can see what's happening now - 

T1 calls dlopen -> g_dl_mutex is locked -> block for our log_access mutex.
T2 calls callback -> locks log_access -> locks config_access -> throws -> unwind happens and dl_iterate_phd   called -> deadlock on g_dl_mutex

We don't see this issue on Windows or Linux - could just be different timings on Android, or maybe unwind is implemented differently.

For now I think it makes sense to do nothing in the context of dlopen and simply call an exported Initialise() method after dlopen.

**Comment by sdaly2107** [comment]
Looks like something similar was changed in GCC7 which explains why we don't see the issue under Linux  - 

https://github.com/scylladb/seastar/issues/73
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=71744
https://gcc.gnu.org/viewcvs/gcc?view=revision&revision=240193

**Comment by DanAlbert** [comment]
The GCC bug there looks likely. For the platform you're seeing the issue on we use libgcc for unwinding, and we never upgraded past GCC 4.9 in Android so we don't have that patch (we'd hoped to be off libgcc by now, but that hasn't happened).

I assume you don't see the bug on 32-bit Arm?

@rprichard since he's been looking in to unwinding stuff.

**Comment by rprichard** [comment]
I saw that gcc patch when I was last looking at exception handling performance. It could improve performance when multiple threads are throwing many exceptions, but it's not that relevant to this issue. The idea is that DSOs can register the location of exception handling information by calling `__register_frame*`, but that's atypical, so the patch added a fast path for skipping the locking/unlocking for the `__register_frame*` state when it's not being used. On Android, I think this fast path would only be relevant for arm64, because our arm32 unwinder doesn't seem to have the `__register_frame*` APIs.

Currently in Android, the loader has a single lock, and both the arm32 and arm64 unwinders acquire this lock to look up EH information for a PC address. They use either `dl_iterate_phdr` (arm64) or `dl_unwind_find_exidx`/`__gnu_Unwind_Find_exidx` (arm32).

I did an experiment with glibc, FreeBSD, and musl:
 - In the executable, dlopen a DSO.
 - In an `__attribute__((constructor))` function in the DSO, spawn, then join a worker thread.
 - In the worker thread, call either `dl_iterate_phdr` or `dlsym`.

With all three C libraries, the worker thread is able to call `dl_iterate_phdr` without deadlocking. FreeBSD and musl also allow the `dlsym` call, whereas glibc deadlocks. `dlsym` can even look up a symbol in the DSO whose constructors are still running. (With Bionic, either API deadlocks.)

Perhaps Bionic needs more finely-grained locking. This CL allows dlsym to run concurrently with constructors, in some cases, but it wouldn't fix this issue: https://android-review.googlesource.com/c/platform/bionic/+/1129355

The [Bugzilla link above](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=71744) has a complaint that `dl_iterate_phdr` doesn't scale to many threads, because only one callback can run concurrently. (There's a suggestion to add a `dl_iterate_phdr_parallel` or `dl_iterate_phdr_rd` function.) I don't know if anyone's actively working on fixing that bug.
